File: leetcode/src/main/java/linked/ConvertBinaryNumberToInteger.java
package linked;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/*-
    https://leetcode.com/problems/convert-binary-number-in-a-linked-list-to-integer/
 */
public class ConvertBinaryNumberToInteger {

    public int getDecimalValue_OK(ListNode head) {
        StringBuilder builder = new StringBuilder();
        while (head != null) {
            builder.append(head.val);
            head = head.next;
        }
        return Integer.parseInt(builder.toString(), 2);
    }

    public int getDecimalValue(ListNode head) {
        StringBuilder builder = new StringBuilder();
        for (ListNode p = head; p != null; p = p.next) {
            builder.append(p.val);
        }
        return Integer.parseInt(builder.toString(), 2);
    }

    @Test
    public void test1() {
        ListNode l = ListNode.of(1, 0, 1);
        assertEquals(5, getDecimalValue(l));
    }

    @Test
    public void test2() {
        ListNode l = ListNode.of(0);
        assertEquals(0, getDecimalValue(l));
    }

}

File: py-algorithm/linked/convert_binary_number_to_integer.py
import unittest
from typing import Optional

import linked.list_node
from linked.list_node import ListNode

"""
    https://leetcode.com/problems/convert-binary-number-in-a-linked-list-to-integer/
"""


class Solution(unittest.TestCase):

    def getDecimalValue(self, head: Optional[ListNode]) -> int:
        result = 0
        while head:
            result = (result << 1) | head.val
            head = head.next
        return result

    def test_1(self):
        head = linked.list_node.ListNode.build_linked_list([1, 0, 1])
        self.assertEqual(self.getDecimalValue(head), 5)

    def test_2(self):
        head = linked.list_node.ListNode.build_linked_list([0])
        self.assertEqual(self.getDecimalValue(head), 0)

    def test_3(self):
        head = linked.list_node.ListNode.build_linked_list([1, 1, 1, 1])
        self.assertEqual(self.getDecimalValue(head), 15)


if __name__ == '__main__':
    unittest.main()
